### ISSUE TITLE
Failed to install stack on Ubuntu 18.04.2

### ISSUE THREAD

**Issue by Pavel-Durov** [commercialhaskell/stack#4816]
I Tried to follow the instructions of stack [documentation](https://docs.haskellstack.org/en/stable/README/) 
Found that stack fails on `setup` step after the installation.

> Ubuntu version: Ubuntu 18.04.2 LTS

### Steps to reproduce
1. Install stack
`$ curl -sSL https://get.haskellstack.org/ | sh `
2. start new project
`$ stack new my-project`
3. navigate to project directory
`$ cd my-project`
4. run setup step
`$ stack setup`

### Expected
`setup` step is completed successfully

### Actual
stack throws an error on `setup` step:
```
$  stack setup --verbose
Version 1.9.3, Git revision 40cf7b37526b86d1676da82167ea8758a854953b (6211 commits) x86_64 hpack-0.31.1
2019-05-11 09:11:45.371763: [debug] Checking for project config at: /home/pavel/nate/haskell/my-project/stack.yaml
2019-05-11 09:11:45.371909: [debug] Loading project config file stack.yaml
2019-05-11 09:11:45.373045: [debug] Decoding build plan from: /home/pavel/.stack/build-plan/lts-13.20.yaml
2019-05-11 09:11:45.373098: [debug] Trying to decode /home/pavel/.stack/build-plan-cache/lts-13.20.cache
2019-05-11 09:11:45.375783: [debug] Success decoding /home/pavel/.stack/build-plan-cache/lts-13.20.cache
2019-05-11 09:11:45.376029: [debug] Run process: /sbin/ldconfig -p
2019-05-11 09:11:45.380010: [debug] Process finished in 4ms: /sbin/ldconfig -p
2019-05-11 09:11:45.380218: [debug] Found shared library libtinfo.so.5 in 'ldconfig -p' output
2019-05-11 09:11:45.380553: [debug] Did not find shared library libtinfo.so.6
2019-05-11 09:11:45.380632: [debug] Did not find shared library libncursesw.so.6
2019-05-11 09:11:45.380687: [debug] Found shared library libgmp.so.10 in 'ldconfig -p' output
2019-05-11 09:11:45.380758: [debug] Did not find shared library libgmp.so.3
2019-05-11 09:11:45.380787: [debug] Potential GHC builds: standard
2019-05-11 09:11:45.380820: [debug] Found already installed GHC builds: standard
2019-05-11 09:11:45.381365: [debug] Performing a sanity check on: /home/pavel/.stack/programs/x86_64-linux/ghc-8.6.5/bin/ghc
2019-05-11 09:11:45.381456: [debug] Run process within /tmp/stack-sanity-check14479/: /home/pavel/.stack/programs/x86_64-linux/ghc-8.6.5/bin/ghc /tmp/stack-sanity-check14479/Main.hs -no-user-package-db
2019-05-11 09:11:45.843495: [debug] Process finished in 462ms: /home/pavel/.stack/programs/x86_64-linux/ghc-8.6.5/bin/ghc /tmp/stack-sanity-check14479/Main.hs -no-user-package-db
The GHC located at /home/pavel/.stack/programs/x86_64-linux/ghc-8.6.5/bin/ghc failed to compile a sanity check. Please see:

    http://docs.haskellstack.org/en/stable/install_and_upgrade/

for more information. Exception was:
Received ExitFailure 1 when running
Raw command: /home/pavel/.stack/programs/x86_64-linux/ghc-8.6.5/bin/ghc /tmp/stack-sanity-check14479/Main.hs -no-user-package-db
Run from: /tmp/stack-sanity-check14479/
Standard output:

[1 of 1] Compiling Main             ( /tmp/stack-sanity-check14479/Main.hs, /tmp/stack-sanity-check14479/Main.o )
Linking /tmp/stack-sanity-check14479/Main ...
Standard error:

/usr/bin/ld.gold: error: cannot find -lgmp
collect2: error: ld returned 1 exit status
`gcc' failed in phase `Linker'. (Exit code: 1)
```

### Stack version

```
$ stack --version
Version 1.9.1, Git revision f9d0042c141660e1d38f797e1d426be4a99b2a3c (6168 commits) x86_64 hpack-0.31.0
```
Looks like it failed to find `libtinfo.so.6/3` dependency?

**Comment by qrilka** [comment]
The error is not about tinfo, try installing `libgmp-dev`

**Comment by Pavel-Durov** [comment]
@qrilka thanks, installing `libgmp-dev` worked!

Should `libgmp-dev` dependency be part of the `curl -sSL https://get.haskellstack.org/ | sh` installation or it is intended to be a separate dependency? 

My question is - is it a bug in the installation script or not? 
I can see that `libgmp-dev` is referenced in the [manual download](https://docs.haskellstack.org/en/stable/install_and_upgrade/#manual-download_2) only :)

**Comment by dbaynard** [comment]
> My question is - is it a bug in the installation script or not?

The installation script probably ought to at least point out that these dependencies are needed. Would you be interested in making a PR, @Pavel-Durov?

**Comment by Pavel-Durov** [comment]
@dbaynard  sure, that would be great, I'll try to find my way :)
Can you point me to where is the installation script located in the codebase?

**Comment by borsboom** [comment]
It's at `etc/scripts/get-stack.sh`.  It should already be installing `libgmp-dev` on Ubuntu, see [line 169](https://github.com/commercialhaskell/stack/blob/master/etc/scripts/get-stack.sh#L169).

**Comment by borsboom** [comment]
Here's what I think may be going on: the command on [line 617](https://github.com/commercialhaskell/stack/blob/master/etc/scripts/get-stack.sh#L617) to check whether any packages need to be installed is:
```
  if dpkg-query -W "$@" > /dev/null 2>&1; then …
```
I believe this turns out to detect if _none_ of the required packages are installed, but doesn't detect if only _some_ of the packages are installed.  So in your case, it may be that one of the other dependencies was already installed (such as `git`) and so it missed installing `libgmp-dev`.

**Comment by borsboom** [comment]
I've opened https://github.com/commercialhaskell/stack/pull/4826 to try to address this.  Please give it a try.

**Comment by Pavel-Durov** [comment]
I was testing `get-stack.sh` script and it worked on plain ubuntu docker image for me. 
here's my docker file:
```
FROM ubuntu:18.10
WORKDIR /user/src/app

RUN apt-get update
RUN apt-get install -y curl
RUN curl -sSL https://get.haskellstack.org/ | sh

ENTRYPOINT stack new my-project && cd my-project && stack setup && stack ghci
```
@borsboom, I tried to install `git` on the same image prior to running `get-stack.sh` and it worked as well 

Dockerfile with git preinstalled:
```
FROM ubuntu:18.10

WORKDIR /user/src/app

RUN apt-get update
RUN apt-get install -y curl
RUN apt-get install -y git

RUN curl -sSL https://get.haskellstack.org/ | sh
ENTRYPOINT stack new my-project && cd my-project && stack setup && stack ghci

```
Any suggestions about how I can replicate the same error as in the ticket?

**Comment by borsboom** [comment]
I've been trying to replicate exactly the error you had (with libgmp-dev) and so far I'm not successful.  I was able to replicate something similar by installing all the dependencies and then purging `netbase`, which put the package in a strange state of being uninstalled but `dpkg-query` still showing it.  The script still thought all the dependencies were installed so didn't reinstall it.  However, removing or purging `libgmp-dev` seems to fully remove the package from the dpkg database, so I didn't run into the same problem there.

In any case, the PR I opened in #4826 should be much better at handling any cases of a package being removed but still in the dpkg database.  I still suspect that's what you ran into, but without a reproduction we can't be certain.
